dev/src/index.ts

```diff
@@ -1198,27 +1218,128 @@ export class Firestore implements firestore.Firestore {
+    const stack = Error().stack!;
+
+    const writer = bulkWriter ?? this.getBulkWriter();
+    writer._verifyNotClosed();
+    const docStream = this._getAllDescendants(
+      ref instanceof CollectionReference
+        ? (ref as CollectionReference<unknown>)
+        : (ref as DocumentReference<unknown>)
+    );
+    const resultDeferred = new Deferred<void>();
+    let errorCount = 0;
+    let lastError: Error | undefined;
+    const incrementErrorCount = (err: Error): void => {
+      errorCount++;
+      lastError = err;
+    };
+    const onStreamEnd = (): void => {
+      if (ref instanceof DocumentReference) {
+        writer.delete(ref).catch(err => incrementErrorCount(err));
+      }
+      writer.flush().then(async () => {
+        if (lastError === undefined) {
+          resultDeferred.resolve();
+        } else {
+          let error = new GoogleError(
+            `${errorCount} ` +
+              `${errorCount !== 1 ? 'deletes' : 'delete'} ` +
+              'failed. The last delete failed with: '
+          );
+          if (lastError instanceof BulkWriterError) {
+            error.code = (lastError.code as number) as Status;
+          }
+          error = wrapError(error, stack);
+
+          // Wrap the BulkWriter error last to provide the full stack trace.
+          resultDeferred.reject(
+            lastError.stack ? wrapError(error, lastError.stack ?? '') : error
+          );
+        }
+      });
+    };
+
+    docStream
+      .on('error', err => {
+        err.code = Status.FAILED_PRECONDITION;
+        err.message = 'Failed to fetch children documents.';
+        lastError = err;
+        onStreamEnd();
+      })
+      .on('data', (snap: QueryDocumentSnapshot) => {
+        writer.delete(snap.ref).catch(err => incrementErrorCount(err));
+      })
+      .on('end', () => onStreamEnd());
+
+    return resultDeferred.promise;
+  }
+
   /**
    * Retrieves all descendant documents nested under the provided reference.
    *
    * @private
    * @return {Stream<QueryDocumentSnapshot>} Stream of descendant documents.
    */
-  // TODO(chenbrian): Make this a private method after adding recursive delete.
-  _getAllDescendants(
-    ref: CollectionReference | DocumentReference
+  private _getAllDescendants(
+    ref: CollectionReference<unknown> | DocumentReference<unknown>
   ): NodeJS.ReadableStream {
     // The parent is the closest ancestor document to the location we're
     // deleting. If we are deleting a document, the parent is the path of that
     // document. If we are deleting a collection, the parent is the path of the
     // document containing that collection (or the database root, if it is a
     // root collection).
+
     let parentPath = ref._resourcePath;
     if (ref instanceof CollectionReference) {
       parentPath = parentPath.popLast();
     }
     const collectionId =
-      ref instanceof CollectionReference ? ref.id : ref.parent.id;
+      ref instanceof CollectionReference
```

Comment:
CollectionReference means `CollectionReference<DocumentData>`. If you change this to `ref instanceof CollectionReference<unknown>` you should be able to remove the `as DocumentReference` cast (which should also be `as DocumentReference<unknown>`).

Reply:
Added the `<unknown>` type to the `DocumentReference` cast.